### PR TITLE
gh-95756: Lazily created cached co_* attrs

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
As long as these aren't accessed, they don't actually consume additional space over 3.11.

* Issue: gh-95756

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
Note that these were all O(1) in 3.10. This restores their behavior. 
https://github.com/python/cpython/blame/08eb754d840696914928355014c2d424131f8835/Objects/codeobject.c

**Comment by markshannon** [comment]
I'd like to reduce the size of code objects for a few reasons. PEP 649 will create lots of small code objects and we want multiple interpreters to share as much data as possible.

File: Doc/c-api/code.rst [hunk 1]
```diff
@@ -93,7 +93,7 @@ bound into a function.
 .. c:function:: PyObject* PyCode_GetVarnames(PyCodeObject *co)
 
    Equivalent to the Python code ``getattr(co, 'co_varnames')``.
-   Returns a new reference to a :c:type:`PyTupleObject` containing the names of
```
Comment (markshannon) [inline Doc/c-api/code.rst]: I think the original form "new reference" is clearer than "strong reference".

File: Doc/c-api/code.rst [hunk 2]
```diff
@@ -102,7 +102,7 @@ bound into a function.
 .. c:function:: PyObject* PyCode_GetCellvars(PyCodeObject *co)
 
    Equivalent to the Python code ``getattr(co, 'co_cellvars')``.
-   Returns a new reference to a :c:type:`PyTupleObject` containing the names of
```
Comment (markshannon) [inline Doc/c-api/code.rst]: ditto

File: Doc/c-api/code.rst [hunk 3]
```diff
@@ -111,7 +111,7 @@ bound into a function.
 .. c:function:: PyObject* PyCode_GetFreevars(PyCodeObject *co)
 
    Equivalent to the Python code ``getattr(co, 'co_freevars')``.
-   Returns a new reference to a :c:type:`PyTupleObject` containing the names of
```
Comment (markshannon) [inline Doc/c-api/code.rst]: ditto

File: Objects/codeobject.c [hunk 4]
```diff
@@ -150,7 +150,22 @@ validate_and_copy_tuple(PyObject *tup)
     return newtuple;
 }
 
+static int
+init_co_cached(PyCodeObject *self) {
+    if (self->_co_cached == NULL) {
+        self->_co_cached = PyMem_New(_PyCoCached, 1);
+        if (self->_co_cached == NULL) {
+            PyErr_NoMemory();
+            return 1;
```
Comment (markshannon) [inline Objects/codeobject.c]: Conventionally `-1` for failure.

File: Objects/codeobject.c [hunk 5]
```diff
@@ -1371,7 +1386,20 @@ _PyCode_SetExtra(PyObject *code, Py_ssize_t index, void *extra)
 PyObject *
 _PyCode_GetVarnames(PyCodeObject *co)
 {
-    return get_localsplus_names(co, CO_FAST_LOCAL, co->co_nlocals);
+    if (init_co_cached(co)) {
+        return NULL;
+    }
+    if (co->_co_cached->_co_varnames != NULL) {
```
Comment (markshannon) [inline Objects/codeobject.c]: You can factor the body of these functions out. `get_cached_locals()` which takes the code object, the flag, and a pointer to the field to be set.

**Review by markshannon — CHANGES_REQUESTED** [review]

File: Objects/codeobject.c [hunk 6]
```diff
@@ -1384,10 +1399,31 @@ _PyCode_SetExtra(PyObject *code, Py_ssize_t index, void *extra)
...
+static PyObject *
+get_cached_locals(PyCodeObject *co, PyObject **cached_field,
+    _PyLocals_Kind kind, int num)
+{
+    assert(cached_field != NULL);
+    assert(co->_co_cached != NULL);
+    if (*cached_field != NULL) {
+        return Py_NewRef(*cached_field);
+    }
+    assert(*cached_field == NULL);
```
Comment (markshannon) [inline Objects/codeobject.c:1411]: No need to assert this. The test two lines above ensures that it is true.

**Review by markshannon — APPROVED** [review]
One assert that could be removed, but it's harmless.

**Comment by Fidget-Spinner** [comment]
I'll leave the assert in since it communicates some info should the code change in the future. And like you mentioned it's harmless anyways.

**Comment by Fidget-Spinner** [comment]
I'm confused. Why is this the only buildbot failing? In any case I will revert this within 24 hours.

**Comment by vstinner** [comment]
> I'm confused. Why is this the only buildbot failing? In any case I might revert this within 24 hours.

I created https://github.com/python/cpython/issues/98178 about this crash. It seems to be specific to macOS and unrelated to your change. If your change would introduce a bug, more tests would crash on more platforms.

**Pull request merged**